### PR TITLE
Add auto device status reporting to nRF Cloud

### PR DESCRIPTION
**Pull request by jayteemo**
Add option to automatically report device status info to nRF Cloud for MQTT devices. This option is enabled by default for devices using FOTA.
Disable this option for ATv2 since it handles this via device messages (pending cloud deployment of IRIS-4689).
Add an app version string to MMSS so it can be reported to nRF Cloud.
IRIS-5235

Update the REST FOTA sample to always send device status info (service info + device info).
Fix project name in REST cell location sample.
Add REST function to send device status as a device message.

### CODE REVIEW

**Comment by jayteemo** [comment]
example data written to shadow for MMSS:
```
"device": {
        "deviceInfo": {
          "appVersion": "1.0.0",
          "modemFirmware": "mfw_nrf9160_1.3.3",
          "batteryVoltage": 5121,
          "imei": "352656109480783",
          "board": "nrf9160dk_nrf9160",
          "sdkVer": "v2.2.99-dev1-32-g3a09fd054e91",
          "appName": "nrf_cloud_mqtt_multi_service",
          "zephyrVer": "v3.2.99-ncs1-1522-g1a5051acc44c",
          "hwVer": "nRF9160 SICA B1A"
        },

```

File: doc/nrf/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -339,6 +339,13 @@ nRF9160 samples
     * MCUboot child image files to properly access external flash on newer nRF9160DK versions.
     * An :file:`overlay_mcuboot_ext_flash.conf` file to enable MCUboot use of external flash.
 
+* :ref:`nrf_cloud_rest_fota` sample:
+
+  * Updated:
+
+    * Device status information, including FOTA enablement, is now sent nRF Cloud when the device connects.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:346]:
```suggestion
    * Device status information, including FOTA enablement, is now sent to nRF Cloud when the device connects.
```

File: doc/nrf/releases/release-notes-changelog.rst [hunk 2]
```diff
@@ -504,6 +511,10 @@ Modem libraries
 
     * Fixed an issue causing the A-GPS data download to be delayed until the RRC connection release.
 
+* :ref:`modem_info_readme` library:
+
+  * Add :c:func:`modem_info_get_hw_version` function to obtain the HW version string using the AT%HWVERSION command.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:516]:
```suggestion
  * Added :c:func:`modem_info_get_hw_version` function to obtain the hardware version string using the ``AT%HWVERSION`` command.
```

File: doc/nrf/releases/release-notes-changelog.rst [hunk 3]
```diff
@@ -540,6 +559,12 @@ Libraries for networking
 
   * Removed location events and event handlers.
 
+* :ref:`lib_nrf_cloud_rest` library:
+
+  * Added:
+
+    * Add the :c:func:`nrf_cloud_rest_device_status_message_send` function to send the device status information as an nRF Cloud device message.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:566]:
```suggestion
    * :c:func:`nrf_cloud_rest_device_status_message_send` function to send the device status information as an nRF Cloud device message.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: samples/nrf9160/nrf_cloud_rest_cell_location/CMakeLists.txt [hunk 4]
```diff
@@ -7,7 +7,7 @@
 cmake_minimum_required(VERSION 3.20.0)
 
 find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
-project(nRF_Cloud_REST_FOTA_Sample)
+project(nRF_Cloud_REST_Cell_Location_Sample)
```
Comment (plskeggs) [inline samples/nrf9160/nrf_cloud_rest_cell_location/CMakeLists.txt:10]: good catch!

**Review by plskeggs — COMMENTED** [review]
LGTM.  As discussed in the Jira ticket, we might need to add the modem UUID, but I'm ok with skipping it for now.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 5]
```diff
@@ -262,6 +235,33 @@ static int get_error_code_value(cJSON *const obj, enum nrf_cloud_error * const e
...
+	if (mdm_inf) {
+		ret = nrf_cloud_modem_info_json_encode(mdm_inf, root_obj);
+		if (ret) {
+			return -ENOMEM;
+		}
+	}
+#endif
+
+	if (svc_inf) {
+		cJSON *svc_inf_obj = cJSON_AddObjectToObjectCS(root_obj, JSON_KEY_SRVC_INFO);
```
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:253]: Could this object creation be moved into `nrf_cloud_service_info_json_encode`?

If you did that, then `info_encode` could be removed entirely, and replaced with just calls to `nrf_cloud_modem_info_json_encode` and `nrf_cloud_service_info_json_encode`

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:253]: Even if not splitting `info_encode`, doing this would increase `info_encode`'s symmetry

**Review by glarsennordic — APPROVED** [review]
LGTM, other than one concern which I won't block over

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 6]
```diff
@@ -547,6 +547,46 @@ int nrf_cloud_encode_config_response(struct nrf_cloud_data const *const input,
 	return 0;
 }
 
+static int add_device_status(cJSON * const reported_obj)
+{
+	/* Only add device status once.
+	 * The application is responsible for keeping dynamic data up to date.
+	 */
+	static bool status_added;
+	int err = 0;
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:556]: i agree.
lets address it in the refactoring epic.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c (hunk 5 above)
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:253]: i'll add this to IRIS-5446

**Review by peknis — APPROVED** [review]

**Comment by plskeggs** [comment]
@jayteemo you seem to have a build failure:
```
/usr/bin/ld: app/libapp.a(nrf_cloud.c.obj): in function `nrf_cloud_init':
/jenkins_workspace/workspace/latest_sdk-nrf_PR-10062/twister-build/native_posix/nrf/tests/subsys/net/lib/nrf_cloud/cloud/net.lib.nrf_cloud.cloud/../../../../../../../../../../nrf/subsys/net/lib/nrf_cloud/src/nrf_cloud.c:115: undefined reference to `nrf_cloud_set_app_version'
/usr/bin/ld: app/libapp.a(nrf_cloud.c.obj): in function `nrf_cloud_shadow_device_status_update':
/jenkins_workspace/workspace/latest_sdk-nrf_PR-10062/twister-build/native_posix/nrf/tests/subsys/net/lib/nrf_cloud/cloud/net.lib.nrf_cloud.cloud/../../../../../../../../../../nrf/subsys/net/lib/nrf_cloud/src/nrf_cloud.c:281: undefined reference to `nrf_cloud_device_status_shadow_encode'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

**Review by tony-le-24 — APPROVED** [review]
LGTM!

File: tests/subsys/net/lib/nrf_cloud/cloud/src/main.c [hunk 7]
```diff
@@ -205,7 +205,7 @@ static void run_before(void *fixture)
 	RESET_FAKE(nct_dc_stream);
 	RESET_FAKE(nct_dc_send);
 	RESET_FAKE(nct_dc_bulk_send);
-	RESET_FAKE(nrf_cloud_device_status_encode);
+	RESET_FAKE(nrf_cloud_device_status_shadow_encode);
```
Comment (plskeggs) [inline tests/subsys/net/lib/nrf_cloud/cloud/src/main.c:208]: You also need to change the 4 other locations in this file that refer to `nrf_cloud_device_status_encode_fake` to be `nrf_cloud_device_status_shadow_encode_fake`.

**Review by plskeggs — CHANGES_REQUESTED** [review]

**Comment by glarsennordic** [comment]
```c
FAKE_VOID_FUNC(nrf_cloud_set_app_version, const char *const);
```
added after line 21 of `tests/subsys/net/lib/nrf_cloud/cloud/src/fakes.h`

Will fix your CI problem. If you @jayteemo or @plskeggs would like me to, I can force-push that change myself

Warning, though, that I'll rebase it into one of the earlier commits, so it could screw up your local repos

**Comment by glarsennordic** [comment]
I have force-pushed the change, so just be aware of that in case you do any further development

**Comment by jayteemo** [comment]
> I have force-pushed the change, so just be aware of that in case you do any further development

:bow:

**Review by plskeggs — APPROVED** [review]

**Pull request merged**